hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/lock/ConditionalWriteLockProvider.java

```diff
@@ -0,0 +1,581 @@
+
+    lockFilePath = buildLockObjectPath(folderName, fileName);
+    ownerId = UUID.randomUUID().toString();
+    this.logger = DEFAULT_LOGGER;
+    this.heartbeatManager = new LockProviderHeartbeatManager(
+        ownerId,
+        heartbeatIntervalMs,
+        this::renewLock);
+
+    try {
+      this.lockService = (StorageLock) ReflectionUtils.loadClass(
+          getLockServiceClassName(uri.getScheme()),
+          new Class<?>[] { String.class, String.class, String.class, Properties.class },
+          new Object[] { ownerId, bucketName, lockFilePath, lockConfiguration.getConfig() });
+    } catch (Throwable e) {
+      throw new HoodieLockException("Failed to load and initialize StorageLock", e);
+    }
+
+    logger.info("Instantiated new Conditional Write LP, owner: {}, lockfilePath: {}", ownerId, lockFilePath);
+  }
+
+  private URI parseURI(String location) {
+    try {
+      return new URI(location);
+    } catch (URISyntaxException e) {
+      throw new HoodieLockException("Unable to parse locks location as a URI: " + location, e);
+    }
+  }
+
+  private static @NotNull String getLockServiceClassName(String scheme) {
+    if (SUPPORTED_LOCK_SERVICE_IMPLEMENTATIONS.containsKey(scheme)
+        && StorageSchemes.isConditionalWriteSupported(scheme)) {
+      return SUPPORTED_LOCK_SERVICE_IMPLEMENTATIONS.get(scheme);
+    } else {
+      throw new HoodieNotSupportedException("No implementation of StorageLock supports this scheme: " + scheme);
+    }
+  }
+
+  @VisibleForTesting
+  ConditionalWriteLockProvider(
+      int heartbeatIntervalMs,
+      int lockValidityMs,
+      String bucketName,
+      String lockFilePath,
+      String ownerId,
+      HeartbeatManager heartbeatManager,
+      StorageLock lockService,
+      Logger logger) {
+    this.heartbeatIntervalMs = heartbeatIntervalMs;
+    this.lockValidityMs = lockValidityMs;
+    this.bucketName = bucketName;
+    this.lockFilePath = lockFilePath;
+    this.heartbeatManager = heartbeatManager;
+    this.lockService = lockService;
+    this.ownerId = ownerId;
+    this.logger = logger;
+    logger.debug("Instantiating new Conditional Write LP, owner: {}", ownerId);
+  }
+
+  // -----------------------------------------
+  // BASE METHODS
+  // -----------------------------------------
+
+  @Override
+  public synchronized StorageLockFile getLock() {
+    return currentLockObj;
+  }
+
+  /**
+   * Attempts to acquire the lock within the given timeout.
+   */
+  @Override
+  public boolean tryLock(long time, TimeUnit unit) {
+    long deadlineNanos = System.nanoTime() + unit.toNanos(time);
+
+    while (System.nanoTime() < deadlineNanos) {
+      try {
+        if (tryLock()) {
+          return true;
+        }
+        logger.debug(
+            LOCK_STATE_LOGGER_MSG,
+            ownerId,
+            lockFilePath,
+            Thread.currentThread(),
+            ACQUIRING);
+        Thread.sleep(DEFAULT_LOCK_ACQUISITION_BUFFER);
```

Comment:
`LockManager#lock` already has retry logic when calling `LockProvider#tryLock`.  Should the retry logic here be removed?

Reply:
We are supposed to tryLock for the appropriate amount of time. I think the implementations in Zookeeper and Dynamo follow the same pattern here

Reply:
> We are supposed to tryLock for the appropriate amount of time
This makes sense, but it does not mean that we should retry at this layer.